app/Models/DuesPackage.php

```diff
@@ -97,30 +101,70 @@ class DuesPackage extends Model
         'access_start' => 'datetime',
         'access_end' => 'datetime',
         'cost' => 'float',
+        'available_for_purchase' => 'boolean',
+        'restricted_to_students' => 'boolean',
     ];
 
     /**
-     * Get the DuesTransaction associated with the DuesPackage model.
+     * Get the DuesTransactions associated with the DuesPackage model.
      */
     public function duesTransactions(): HasMany
     {
         return $this->hasMany(DuesTransaction::class);
     }
 
     /**
-     * Get the DuesTransaction associated with the DuesPackage model.
+     * Get the DuesTransactions associated with the DuesPackage model.
      */
     public function transactions(): HasMany
     {
         return $this->hasMany(DuesTransaction::class);
     }
 
+    /**
+     * Get the FiscalYear associated with the DuesPackage model.
+     */
+    public function fiscalYear(): BelongsTo
+    {
+        return $this->belongsTo(FiscalYear::class);
+    }
+
+    public function conflictsWith(): BelongsTo
+    {
+        return $this->belongsTo(self::class, 'conflicts_with_package_id');
+    }
+
+    public function hasConflictWith(): HasMany
+    {
+        return $this->hasMany(self::class, 'conflicts_with_package_id');
+    }
+
     /**
      * Scope a query to only include DuesPackages available for purchase.
      */
     public function scopeAvailableForPurchase(Builder $query): Builder
     {
-        return $query->where('available_for_purchase', 1);
+        return $query->where('available_for_purchase', true);
+    }
+
+    public function scopeUserCanPurchase(Builder $query, User $user): Builder
+    {
+        return $query
+            ->select('dues_packages.*')
+            ->leftJoin('dues_transactions', static function (JoinClause $join) use ($user): void {
+                $join->on('dues_packages.conflicts_with_package_id', '=', 'dues_transactions.dues_package_id')
+                     ->where('dues_transactions.user_id', $user->id);
+            })
+            ->leftJoin('payments', static function (JoinClause $join): void {
+                $join->on('payments.payable_id', '=', 'dues_transactions.id')
+                        ->where('payments.payable_type', '=', DuesTransaction::getMorphClassStatic())
+                        ->where('payments.deleted_at', '=', null)
+                        ->where('payments.amount', '>', 0);
+            })
+            ->where('available_for_purchase', true)
+            ->where('dues_packages.effective_end', '>=', date('Y-m-d'))
+            ->where('restricted_to_students', 'student' === $user->primary_affiliation)
```

Comment:
are we masking `credit-student` into `student`?

Reply:
We don't see `credit-student`